nixos/tests/binary-cache.nix

```diff
@@ -0,0 +1,62 @@
+import ./make-test-python.nix ({ lib, ... }:
+
+with lib;
+
+{
+  name = "binary-cache";
+  meta.maintainers = with maintainers; [ thomasjm ];
+
+  nodes.machine =
+    { pkgs, ... }: {
+      imports = [ ../modules/installer/cd-dvd/channel.nix ];
+      environment.systemPackages = with pkgs; [python3];
+      system.extraDependencies = with pkgs; [hello.inputDerivation];
+      nix.extraOptions = ''
+        experimental-features = nix-command
+      '';
+    };
+
+  testScript = ''
+    # Build the cache, then remove it from the store
+    cachePath = machine.succeed("nix-build --no-out-link -E 'with import <nixpkgs> {}; mkBinaryCache { rootPaths = [hello]; }'").strip()
+    machine.succeed("cp -r %s/. /tmp/cache" % cachePath)
+    machine.succeed("nix-store --delete " + cachePath)
+
+    # Sanity test of cache structure
+    status, stdout = machine.execute("ls /tmp/cache")
+    cache_files = stdout.split()
+    assert ("nix-cache-info" in cache_files)
+    assert ("nar" in cache_files)
+
+    # Nix store ping should work
+    machine.succeed("nix store ping --store file:///tmp/cache")
+
+    # Cache should contain a .narinfo referring to "hello"
+    grepLogs = machine.succeed("grep -l 'StorePath: /nix/store/[[:alnum:]]*-hello-.*' /tmp/cache/*.narinfo")
+
+    # Get the store path referenced by the .narinfo
+    narInfoFile = grepLogs.strip()
+    narInfoContents = machine.succeed("cat " + narInfoFile)
+    import re
+    match = re.match(r"^StorePath: (/nix/store/[a-z0-9]*-hello-.*)$", narInfoContents, re.MULTILINE)
+    if not match: raise Exception("Couldn't find hello store path in cache")
+    storePath = match[1]
+
+    # Delete the store path
+    machine.succeed("nix-store --delete " + storePath)
+    machine.succeed("[ ! -d %s ] || exit 1" % storePath)
+
```

Comment:
```suggestion

    # Should not be able to build hello without the cache, to check the effectiveness of the test
    machine.fail("nix-build -A hello '<nixpkgs>' --option require-sigs false 2>&1")


```

Reply:
@roberth the above doesn't make sense as written. This VM is set up with all the dependencies of `hello`, so it's expected that it will be able to build it. (Just as it does when building it in order to put it in the cache.)

Reply:
I suppose you're right and the log grepping below makes up for this fact a bit, but I would rather not rely on that, because when it's not necessary to use the cache, some errors may be recovered from.

A simple test is better than a complicated one, and a setup where this assertion holds up would be simpler. I guess such simplification may not be feasible without 
 - https://github.com/NixOS/nix/pull/7087

I've tried to hide the references with an intermediate tar.gz derivation, but Nix was clever enough to rediscover them in my subsequent unpacking derivation.

Reply:
Maybe a way to ensure that `hello` isn't being built from scratch when using the cache is to remove the input derivation stuff from the Nix store after the cache is built. But that goes back to the question I had before about how you can remove such store path(s) given they're GC roots in the NixOS environment.

I think the log grepping is a pretty good test, the only thing I feel a little bad about is that it may be brittle if log output changes.